File: lib_algorithms_sort/src/main/java/net/chris/lib/algorithms/sort/SelectionSorter.java
package net.chris.lib.algorithms.sort;

/**
 * Selection sort.
 *
 * @param <T>
 */
public abstract class SelectionSorter<T> extends Sorter<T> {

    @Override
    protected void subSort(T[] array) {
        if (array == null) {
            return;
        }
        for (int i = 0; i < array.length - 1; i++) {
            int min = i;
            T temp;
            //find min
            for (int j = i + 1; j < array.length; j++) {
                if (compareTo(array[j], array[min]) < 0) {
                    min = j;
                }
            }
            //swap the min with the ith element
            temp = array[min];
            array[min] = array[i];
            array[i] = temp;
            doUpdate(array);
        }
    }

}

File: lib_algorithms_sort/src/main/cpp/SelectionSorter.cpp
#include "SelectionSorter.h"

void selectSort(int *array, int size) {
    int temp;
    int pos;
    for (int i = 0; i < size - 1; i++) {
        temp = array[i];
        pos = i;
        for (int j = i + 1; j < size; j++) {
            if (array[j] < temp) {
                temp = array[j];
                pos = j;
            }
        }
        array[pos] = array[i];
        array[i] = temp;
    }
}